dipy/tracking/fast_tracking.pyx

```diff
@@ -0,0 +1,256 @@
+# cython: boundscheck=False
+# cython: initializedcheck=False
+# cython: wraparound=False
+# cython: Nonecheck=False
+
+from libc.stdio cimport printf
+
+cimport cython
+from cython.parallel import prange
+import numpy as np
+cimport numpy as cnp
+
+from dipy.core.interpolation cimport trilinear_interpolate4d_c
+from dipy.direction.pmf cimport PmfGen
+from dipy.tracking.stopping_criterion cimport StoppingCriterion
+from dipy.utils cimport fast_numpy
+
+from dipy.tracking.stopping_criterion cimport (StreamlineStatus,
+                                               StoppingCriterion,
+                                               TRACKPOINT,
+                                               ENDPOINT,
+                                               OUTSIDEIMAGE,
+                                               INVALIDPOINT,
+                                               VALIDSTREAMLIME,
+                                               INVALIDSTREAMLIME)
+from dipy.tracking.tracker_parameters cimport (TrackerParameters, 
+                                               func_ptr, 
+                                               SUCCESS, 
+                                               FAIL)
+
+from nibabel.streamlines import ArraySequence as Streamlines
+
+from libc.stdlib cimport malloc, free
+from libc.string cimport memcpy
+from libc.math cimport floor, ceil
+from libc.stdio cimport printf
+
+from posix.time cimport clock_gettime, timespec, CLOCK_REALTIME
+
+cdef extern from "stdlib.h" nogil:
+    void *memset(void *ptr, int value, size_t num)
+
+
+def generate_tractogram(double[:,::1] seed_positions,
+                        double[:,::1] seed_directions,
+                        StoppingCriterion sc,
+                        TrackerParameters params,
+                        PmfGen pmf_gen,
+                        int nbr_threads=0,
+                        float buffer_frac=1.0):
+    """
+    return_all : bool, optional
+        If true, return all generated streamlines, otherwise only
+        streamlines reaching end points or exiting the image.
+    """
+
+    cdef:
+        cnp.npy_intp _len = seed_positions.shape[0]
+        cnp.npy_intp _plen = int(ceil(_len * buffer_frac))
+        cnp.npy_intp i, seed_start, seed_end
+        double** streamlines_arr
+        int* length_arr
+        StreamlineStatus* status_arr
+
+    # srand/rand don't play well with multi-threading
+    if params.random_seed > 0 and not nbr_threads == 1:
+        raise ValueError("random_seed > 0 do not work with nbr_threads != 1.")
```

Comment:
Does this still make sense. Especially when you look later in the code this (line 157): 

```
    if params.random_seed > 0:
        s_random_seed = int(
            (seed[0] * 2 + seed[1] * 3 + seed[2] * 5) * params.random_seed
            )
    else:
        clock_gettime(CLOCK_REALTIME, &ts)
        s_random_seed = int(ts.tv_sec + (ts.tv_nsec / 1000000000.))
    fast_numpy.seed(s_random_seed)
```